### ISSUE TITLE
Consider dropping the VM number length limit

### ISSUE THREAD

**Issue by A60AB5450353F40E** [bitjson/bch-bigint#1]
This would allow [implementing RSA-4096](https://x.com/reardencode/status/1823456943392088119)

---

Maintainer note by @bitjson: I renamed this issue from `Consider 520 bytes` to `Consider dropping the VM number length limit`.

**Comment by bitjson** [comment]
Thanks for opening the issue @A60AB5450353F40E!

I would definitely like this CHIP to evaluate a higher limit. I wanted to initially propose 258 bytes as a conservative minimum, but **I have yet to identify any reason(s) for retaining the number length limit at all**. (In some ways, removing the limit altogether would be more conservative in that it would ensure we don't have to mess with it again – like ABLA with block sizes.)

I'll add there's also no reason we'd single out 520 bytes for a limit either – if the [Limits CHIP](https://github.com/bitjson/bch-vm-limits) is activated, the BCH protocol will no longer have any constant(s) set to 520 (as the push limit would move up to the same 10,000 byte constant as `MAX_SCRIPT_SIZE`).

Without an explicit limit, the "maximum number size" simply becomes the 10,000 byte push limit for most operations, and somewhere around 2800 bytes (`sqrt(MAX_SCRIPT_SIZE*800)`) for `OP_MUL`, `OP_DIV`, and `OP_MOD` assuming only one operation in a max-length contract. (That estimate may still change as more node implementations review the limits CHIP.)

A non-obvious protocol complexity upside of completely removing this limit: there are far fewer edge cases to verify across all number accepting operations – right now, correct implementations need to carefully fail in all the right circumstances given various combinations of padded (non-minimal), negative, padded-negative, out-of-range, etc. numbers (and especially `OP_NUM2BIN` and `OP_BIN2NUM` are easy to mis-implement). Cutting out another unessential limit reduces that combinatorial test set by quite a lot.

In other words, with the limits CHIP, implementations would already be required to perfectly implement and test arithmetic operation cost limits; **"leaving in" a separate VM number length limit is sort of arbitrary, has no impact on worst-case performance, might be a hassle for some use cases, and could create demand for otherwise-unnecessary future increase(s)**.

**Comment by jsmith-dev** [comment]
+1 for arbitrary limit
